src/Microsoft.ML.Tokenizers/Utils/StringSpanOrdinalKey.cs

```diff
@@ -0,0 +1,131 @@
+            Span.SequenceEqual(other.Span);
+
+        public override int GetHashCode() => Helpers.GetHashCode(Span);
+    }
+
+    internal sealed class StringSpanOrdinalKeyCache<TValue>
+    {
+        private readonly int _capacity;
+        private readonly Dictionary<StringSpanOrdinalKey, TValue> _map;
+
+        private object SyncObj => _map;
+
+        internal StringSpanOrdinalKeyCache() : this(Bpe.DefaultCacheCapacity) { }
+
+        internal StringSpanOrdinalKeyCache(int capacity)
+        {
+            _capacity = capacity;
+            _map = new Dictionary<StringSpanOrdinalKey, TValue>(capacity);
+        }
+
+        internal bool TryGetValue(string key, out TValue value)
+        {
+            lock (SyncObj)
+            {
+                return _map.TryGetValue(new StringSpanOrdinalKey(key), out value!);
+            }
+        }
+
+        internal unsafe bool TryGetValue(ReadOnlySpan<char> key, out TValue value)
+        {
+            lock (SyncObj)
+            {
+                fixed (char* ptr = key)
+                {
+                    return _map.TryGetValue(new StringSpanOrdinalKey(ptr, key.Length), out value!);
+                }
+            }
+        }
+
+        internal void Remove(string key)
+        {
+            lock (SyncObj)
+            {
+                _map.Remove(new StringSpanOrdinalKey(key));
+            }
+        }
+
+        internal void Set(string k, TValue v)
+        {
+            lock (SyncObj)
+            {
+                if (_map.Count < _capacity)
+                {
+                    _map[new StringSpanOrdinalKey(k)] = v;
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Custom JSON converter for <see cref="StringSpanOrdinalKey"/>.
+    /// </summary>
+    internal sealed class StringSpanOrdinalKeyConverter : JsonConverter<StringSpanOrdinalKey>
+    {
+        public override StringSpanOrdinalKey ReadAsPropertyName(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) =>
+            new StringSpanOrdinalKey(reader.GetString()!);
+
+        public override void WriteAsPropertyName(Utf8JsonWriter writer, StringSpanOrdinalKey value, JsonSerializerOptions options) =>
+            writer.WriteStringValue(value.Data!);
+
+        public override StringSpanOrdinalKey Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) => new StringSpanOrdinalKey(reader.GetString()!);
+        public override void Write(Utf8JsonWriter writer, StringSpanOrdinalKey value, JsonSerializerOptions options) => writer.WriteStringValue(value.Data!);
+    }
+
+    /// <summary>
+    /// Extension methods for <see cref="StringSpanOrdinalKey"/>.
+    /// </summary>
+    internal static class StringSpanOrdinalKeyExtensions
+    {
+        public unsafe static bool TryGetValueUnsafe<TValue>(this IReadOnlyDictionary<StringSpanOrdinalKey, TValue> map, ReadOnlySpan<char> key, out TValue value)
```

Comment:
Why is this "Unsafe"?

Also, it's an extension on IReadOnlyDictionary rather than Dictionary. Is that necessary, or is everywhere it's used on a Dictionary?

Reply:
Addressed both issues. Thanks!